World builder separate cmake project (#3771)
geodynamics/aspect: CMakeLists.txt

DEAL_II_INITIALIZE_CACHED_VARIABLES()
PROJECT(${TARGET} CXX)

> include(${CMAKE_ROOT}/Modules/ExternalProject.cmake)

Reviewer: ```suggestion
INCLUDE(${CMAKE_ROOT}/Modules/ExternalProject.cmake)
```

    SET(WORLD_BUILDER_SOURCE_DIR "${CMAKE_SOURCE_DIR}/contrib/world_builder/" CACHE PATH "" FORCE)
  ENDIF()

  SET(WORLD_BUILDER_INSTALL_DIR ${CMAKE_BINARY_DIR}/world_builder_${CMAKE_BUILD_TYPE}/)
  SET(WB_LIB_LOCATION ${WORLD_BUILDER_INSTALL_DIR}lib/${CMAKE_STATIC_LIBRARY_PREFIX}WorldBuilder${CMAKE_STATIC_LIBRARY_SUFFIX})

  ExternalProject_Add(world_builder_project
  SOURCE_DIR ${WORLD_BUILDER_SOURCE_DIR}
  TMP_DIR ${WORLD_BUILDER_INSTALL_DIR}tmp/
  BINARY_DIR ${WORLD_BUILDER_INSTALL_DIR}
  INSTALL_COMMAND ""
  INSTALL_DIR ${WORLD_BUILDER_INSTALL_DIR}
  BUILD_BYPRODUCTS ${WB_LIB_LOCATION}
  CMAKE_ARGS -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} -DCMAKE_INSTALL_PREFIX=${WORLD_BUILDER_INSTALL_DIR} ${WORLD_BUILDER_SOURCE_DIR}
)

ExternalProject_Get_Property(world_builder_project BINARY_DIR)

> add_library(WorldBuilder STATIC IMPORTED)

Author: by default it makes a static library. In the future I can make it a cmake option, but I don't see why a dynamic library would be better in our case.

set_target_properties(WorldBuilder PROPERTIES

>   IMPORTED_LOCATION ${WB_LIB_LOCATION})

Author: It tells cmake where to find the world builder lib when it is build

    SET(WORLD_BUILDER_SOURCE_DIR "${CMAKE_SOURCE_DIR}/contrib/world_builder/" CACHE PATH "" FORCE)
  ENDIF()

>   string(TOLOWER ${CMAKE_BUILD_TYPE} BUILD_TYPE_LOWER)

Reviewer: That last argument here looks like it's a built-in cmake something because you spell it all upper-case. But it's really just the name of a local variable.

There is an informal cmake convention that temporary variables are spelled as `_temp_variable`. So I would suggest using
```suggestion
  STRING(TOLOWER ${CMAKE_BUILD_TYPE} _build_type_lower)
```

  SET(WORLD_BUILDER_INSTALL_DIR ${CMAKE_BINARY_DIR}/world_builder_${BUILD_TYPE_LOWER}/)

>   SET(WB_LIB_LOCATION ${WORLD_BUILDER_INSTALL_DIR}lib/${CMAKE_SHARED_LIBRARY_PREFIX}WorldBuilder${CMAKE_SHARED_LIBRARY_SUFFIX})

Reviewer: the same applies to the two variables you set here

  ExternalProject_Add(world_builder_project

>   SOURCE_DIR ${WORLD_BUILDER_SOURCE_DIR}

Reviewer: Can you align the arguments here as we do in many other places?

ADD_EXECUTABLE(${TARGET} ${TARGET_SRC})
DEAL_II_SETUP_TARGET(${TARGET})

if(ASPECT_WITH_WORLD_BUILDER)
  # Make sure that the whole library is loaded, so the registration is done correctly.

>   if(NOT APPLE AND NOT MSVC)

Author: The reason why this is needed is that the world builder has a plugin system similar to what aspect has. Which would normally be fine, but  If this is not set, the plugins are not included. Unfortunately, each platform has their own way of doing this and I don't see a way around it if we want to have the world builder as a library. But if you know how to do it, let me know :)

>   if(NOT APPLE AND NOT MSVC)

Reviewer 2: Are you compiling the plugins into the same lib or separate libs?

>   if(NOT APPLE AND NOT MSVC)

Author: the same lib

    SET(GWB_LIBRARY_WHOLE -Wl,--whole-archive WorldBuilder -Wl,--no-whole-archive)
  elseif(MSVC)
    SET(GWB_LIBRARY_WHOLE WorldBuilder)
  else()
    SET(GWB_LIBRARY_WHOLE -Wl,-force_load WorldBuilder)
  endif()

  target_link_libraries(${TARGET} ${GWB_LIBRARY_WHOLE})

  # Make doesn't know how to how to layout the dependencies
  # in a fine grained manner and requires a more coarse
  # approach than Ninja. Since I don't know whether this also
  # works for other build systems, the caorse approach is only
  # disabled for Ninja.
  if(NOT ${CMAKE_GENERATOR} STREQUAL "Ninja" )

>     add_dependencies( WorldBuilder world_builder_project )

Author: Yes, but this make ninja faster (on my machine with 6 cores, 12 threads, it is 1m52s vs 2m7s, I tried it several times and the compile times are quite stable).

>     add_dependencies( WorldBuilder world_builder_project )

Reviewer 2: I don't understand why that is the case. Can you explain?

>     add_dependencies( WorldBuilder world_builder_project )

Author: If you add the statment you force ninja to only start building aspect when the world builder is done. In practise that is a to restrictive policy. It can actually start to build everything in a smart way without waiting for the world builder to be compiled, and only waiting for the things which actually need to wait. The more parallel your build is, more more it will save you time because it doesn't have to wait.

>     add_dependencies( WorldBuilder world_builder_project )

Author: btw, here is the ninja ouput where you can see (even though it is in parrallel) that ninja is building stuff in an intelegent way:

```
[1/48] Creating directories for 'world_builder_project'
[2/48] No download step for 'world_builder_project'
[3/48] No update step for 'world_builder_project'
[4/48] No patch step for 'world_builder_project'
[5/48] Performing configure step for 'world_builder_project'
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Git: /usr/bin/git (found version "2.25.1") 
-- Building the Geodynamic World Builder version 0.3.0 in Debug mode
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - /usr/bin/f95
-- The Fortran compiler identification is GNU 9.3.0
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/f95 - skipped
-- Checking whether /usr/bin/f95 supports Fortran 90
-- Checking whether /usr/bin/f95 supports Fortran 90 - yes
-- Found Fortran support. Enabling Fortran wrapper and tests.
-- Found SWIG: /usr/bin/swig4.0 (found version "4.0.1")  
-- Found Python: /usr/bin/python3.8 (found version "3.8.2") found components: Interpreter Development Development.Module Development.Embed 
-- Both SWIG and Python are found, enableling python wrapper compilation.
-- Performing Test SUGGEST_OVERWRITE_CXX_COMPILE_FLAG
-- Performing Test SUGGEST_OVERWRITE_CXX_COMPILE_FLAG - Success
-- Performing Test NO_PLACEMENT_NEW_CXX_COMPILE_FLAG
-- Performing Test NO_PLACEMENT_NEW_CXX_COMPILE_FLAG - Success
-- Found /usr/bin/numdiff to test the difference between files.
-- Configuring done
-- Generating done
-- Build files have been written to: [...]/aspect/aspect-C/build-dg-u-nj/world_builder_Debug
[6/48] Building CXX object CMakeFiles/aspect.dir/cmake_pch.hxx.gch
[7/48] Performing build step for 'world_builder_project'
[1/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/config.cc.o
[2/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/composition/uniform.cc.o
[3/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/coordinate_systems/spherical.cc.o
[4/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/coordinate_systems/cartesian.cc.o
[5/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/temperature/adiabatic.cc.o
[6/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/temperature/interface.cc.o
[7/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/composition/interface.cc.o
[8/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/temperature/linear.cc.o
[9/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/coordinate_systems/interface.cc.o
[10/89] Building CXX object CMakeFiles/WorldBuilderApp.dir/app/main.cc.o
[11/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate.cc.o
[12/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/grains/interface.cc.o
[13/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/grains/random_uniform_distribution.cc.o
[14/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/grains/uniform.cc.o
[15/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/continental_plate_models/temperature/uniform.cc.o
[16/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/composition/uniform.cc.o
[17/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/composition/interface.cc.o
[18/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/temperature/adiabatic.cc.o
[19/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/grains/interface.cc.o
[20/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/temperature/linear.cc.o
[21/89] Building CXX object CMakeFiles/WorldBuilderVisualization.dir/visualization/main.cc.o
[22/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/grains/random_uniform_distribution.cc.o
[23/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/temperature/interface.cc.o
[24/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/grains/uniform.cc.o
[25/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault_models/temperature/uniform.cc.o
[26/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer.cc.o
[27/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/composition/interface.cc.o
[28/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/temperature/adiabatic.cc.o
[29/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/composition/uniform.cc.o
[30/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/grains/interface.cc.o
[31/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/grains/uniform.cc.o
[32/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/fault.cc.o
[33/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/temperature/linear.cc.o
[34/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/temperature/uniform.cc.o
[35/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/temperature/interface.cc.o
[36/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/composition/uniform.cc.o
[37/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/interface.cc.o
[38/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/mantle_layer_models/grains/random_uniform_distribution.cc.o
[39/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/composition/interface.cc.o
[40/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/grains/interface.cc.o
[41/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate.cc.o
[42/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/temperature/linear.cc.o
[43/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/grains/random_uniform_distribution.cc.o
[44/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/grains/uniform.cc.o
[45/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/temperature/adiabatic.cc.o
[46/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/temperature/interface.cc.o
[47/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/temperature/uniform.cc.o
[48/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/oceanic_plate_models/temperature/plate_model.cc.o
[49/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/composition/uniform.cc.o
[50/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/grains/uniform.cc.o
[51/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/composition/interface.cc.o
[52/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/grains/random_uniform_distribution.cc.o
[53/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/utilities.cc.o
[54/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/temperature/adiabatic.cc.o
[55/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/grains/interface.cc.o
[56/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/point.cc.o
[57/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/temperature/uniform.cc.o
[58/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/temperature/plate_model.cc.o
[59/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/temperature/linear.cc.o
[60/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate_models/temperature/interface.cc.o
[61/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/array.cc.o
[62/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/interface.cc.o
[63/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/plugin_system.cc.o
[64/89] Building Fortran preprocessed CMakeFiles/WorldBuilder.dir/source/wrapper_fortran.f90-pp.f90
[65/89] Generating Fortran dyndep file CMakeFiles/WorldBuilder.dir/Fortran.dd
[66/89] Building Fortran object CMakeFiles/WorldBuilder.dir/source/wrapper_fortran.f90.o
[67/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/double.cc.o
[68/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/bool.cc.o
[69/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/object.cc.o
[70/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/string.cc.o
[71/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/features/subducting_plate.cc.o
[72/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/point.cc.o
[73/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/unsigned_int.cc.o
[74/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/types/segment.cc.o
[75/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/wrapper_cpp.cc.o
[76/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/wrapper_c.cc.o
[77/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/world.cc.o
[78/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/utilities.cc.o
[79/89] Building CXX object CMakeFiles/WorldBuilder.dir/source/parameters.cc.o
[80/89] Linking CXX static library lib/libWorldBuilder.a
[81/89] Swig compile [...]/aspect/aspect-C/contrib/world_builder/source/wrapper_cpp.i for python
[82/89] Building CXX object CMakeFiles/gwb.dir/lib/wrapper_cppPYTHON_wrap.cxx.o
[83/89] Linking CXX executable bin/WorldBuilderVisualization
[84/89] Linking CXX executable bin/WorldBuilderApp
[85/89] Linking CXX shared module lib/_gwb.so
[86/89] Building CXX object tests/CMakeFiles/unit_test_wb_glm.dir/unit_tests/unit_test_wb_glm.cc.o
[87/89] Linking CXX executable bin/unit_test_wb_glm
[88/89] Building CXX object tests/CMakeFiles/unit_test_world_builder.dir/unit_tests/unit_test_world_builder.cc.o
[89/89] Linking CXX executable bin/unit_test_world_builder
[8/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_31_cxx.cxx.o
[9/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_30_cxx.cxx.o
[10/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_29_cxx.cxx.o
[11/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_28_cxx.cxx.o
[12/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_36_cxx.cxx.o
[13/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_26_cxx.cxx.o
[14/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_27_cxx.cxx.o
[15/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_25_cxx.cxx.o
[16/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_35_cxx.cxx.o
[17/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_21_cxx.cxx.o
[18/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_20_cxx.cxx.o
[19/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_19_cxx.cxx.o
[20/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_17_cxx.cxx.o
[21/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_18_cxx.cxx.o
[22/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_23_cxx.cxx.o
[23/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_24_cxx.cxx.o
[24/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_14_cxx.cxx.o
[25/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_13_cxx.cxx.o
[26/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_22_cxx.cxx.o
[27/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_15_cxx.cxx.o
[28/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_9_cxx.cxx.o
[29/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_33_cxx.cxx.o
[30/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_7_cxx.cxx.o
[31/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_10_cxx.cxx.o
[32/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_12_cxx.cxx.o
[33/48] No install step for 'world_builder_project'
[34/48] Completed 'world_builder_project'
[35/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_8_cxx.cxx.o
[36/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_32_cxx.cxx.o
[37/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_6_cxx.cxx.o
[38/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_37_cxx.cxx.o
[39/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_5_cxx.cxx.o
[40/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_0_cxx.cxx.o
[41/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_3_cxx.cxx.o
[42/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_1_cxx.cxx.o
[43/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_4_cxx.cxx.o
[44/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_16_cxx.cxx.o
[45/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_2_cxx.cxx.o
[46/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_11_cxx.cxx.o
[47/48] Building CXX object CMakeFiles/aspect.dir/Unity/unity_34_cxx.cxx.o
[48/48] Linking CXX executable aspect
```

>     add_dependencies( WorldBuilder world_builder_project )

Reviewer: I think you should add the dependency on the aspect executable, not the aspect project.

ADD_EXECUTABLE(${TARGET} ${TARGET_SRC})
DEAL_II_SETUP_TARGET(${TARGET})

if(ASPECT_WITH_WORLD_BUILDER)
  target_link_libraries(${TARGET} WorldBuilder)

>   # Make doesn't know how to how to layout the dependencies

Reviewer: ```suggestion
  # Make doesn't know how to lay out the dependencies
```
